File: Python/Minimum Hours of Training to Win a Competition.py
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def minNumberOfHours(self, energy: int, experience: int, energies: List[int], experiences: List[int]) -> int:
        hours = 0

        for eng, exp in zip(energies, experiences):
            # Adding the missing amount of energy
            extra_en = max(0, eng - energy + 1)
            energy += extra_en

            # Adding the missing amount of experience
            extra_ex = max(0, exp - experience + 1)
            experience += extra_ex

            energy -= eng
            experience += exp
            hours += extra_en + extra_ex

        return hours

File: Java/Minimum Hours of Training to Win a Competition.java
public class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public int minNumberOfHours(int initialEnergy, int initialExperience, int[] energy, int[] experience) {
        int hours = 0;

        for (int i = 0; i < energy.length; ++i) {
            // Adding the missing amount of energy
            int extraEnergy = Math.max(0, energy[i] - initialEnergy + 1);
            initialEnergy += extraEnergy;

            // Adding the missing amount of experience
            int extraExperience = Math.max(0, experience[i] - initialExperience + 1);
            initialExperience += extraExperience;

            initialEnergy -= energy[i];
            initialExperience += experience[i];

            hours += extraEnergy + extraExperience;
        }

        return hours;
    }
}
